### ISSUE TITLE
Multiple opatch commands not working successfully

### ISSUE THREAD

**Issue by rrod157** [biemond/biemond-orawls#451]
I have 4 orawls::opatch commands because i need to apply 4 different patches in my install. But somehow it randomly fails on some patches. Here is a sample log output:

[ Error during Oracle Home discovery Phase]. Detail: OPatchSession cannot load inventory for the given Oracle Home /appl/ohs12c/Oracle/Middleware/Oracle_Home. Possible causes are:
   No read or write permission to ORACLE_HOME/.patch_storage
   Central Inventory is locked by another OUI instance
   No read permission to Central Inventory
   The lock file exists in ORACLE_HOME/.patch_storage
   The Oracle Home does not exist in Central Inventory

[ Error during Oracle Home discovery Phase]. Detail: OPatch failed: ApplySession failed to prepare the system. com/oracle/cie/gdr/dao/pch/BugList

Recommended actions: Please make sure no other OPatch or OUI processes is running. Try running $ORACLE_HOME/oui/bin/runInstsaller.

OPatch failed with error code 22

Although it patches the missing patch the next time it run, we are expecting that it shouldn't fail because of overlapping opatch commands. Perhaps there is a missing dependency that might prevent this to fail?  like some kind of a marker checking or something??

**Comment by biemond** [comment]
Hi,

For db and grid I am also doing this ( basically the same code) but these are mostly big patches. 
Maybe it goes a little bit too fast and something is still running in the background.
Maybe we need to add a sleep of a few seconds in between.

thanks Edwin

**Comment by rrod157** [comment]
is there a workaround on this one? although it applies the patch the next time it runs, but the system is already up. Some patch cannot be applied if the system is running:

Is the local system ready for patching? [y|n]
Y (auto-answered by -silent)
User Responded with: Y
Backing up files...

Patching component oracle.ohs2, 12.1.3.0.0...

Patching component oracle.ohs2, 12.1.3.0.0...

Copy failed from '/tmp/puppet_binary/wls_opatch/24329181/files/oracle.ohs2/12.1.3.0.0/oracle.ohs2.symbol/bin/httpd.worker' to '/appl/ohs12c/Oracle/Middleware/Oracle_Home/ohs/bin/httpd.worker'...
Please verify all applications associated with the Oracle Home '/appl/ohs12c/Oracle/Middleware/Oracle_Home' are shut down. If this is AIX, please perform solution documented in Note 739963.1 on https://myoraclesupport.oracle.com.

**Comment by rrod157** [comment]
I think you can update the underlying process to check the lock file inside ORACLE_HOME/.patch_storage

The error log produce this info:
The lock file exists in ORACLE_HOME/.patch_storage

If we can just subscribe to that lock file, maybe we can prevent (wait) the next patch to run?

**Comment by biemond** [comment]
so we should wait after applying the patch until this file is gone , with some timeout of 10 seconds

**Comment by biemond** [comment]
do it here 
https://github.com/biemond/biemond-orawls/blob/puppet4/lib/puppet/provider/opatch/opatch.rb#L40

or wait before retrieving the status
https://github.com/biemond/biemond-orawls/blob/puppet4/lib/puppet/provider/opatch/opatch.rb#L54

**Comment by rrod157** [comment]
I think it is safe to add a sleep of 10 seconds. Can we put this in L54?
sleep(10)

**Comment by biemond** [comment]
let's add add that before the fail  https://github.com/biemond/biemond-orawls/blob/puppet4/lib/puppet/provider/opatch/opatch.rb#L40

else you have to wait even if you don't do a patch or already applied that one.

Go ahead and let me know if it fixes for you.

**Comment by rrod157** [comment]
Should i add it directly to the code in our puppet master? it is not in our github repo but i can see the file inside the puppet master.

~># cat opatch.rb | wc -l
93
~>#

**Comment by biemond** [comment]
yep, maybe also add an info log, do a restart of the pupper master and do a puppet run on the master so it will pickup this change

**Comment by userMar2017** [comment]
Is there anyway to pass "-delay" and "-retry" to opatch using your method ?

**Comment by biemond** [comment]
are these opatch parameter you do want to implement them on the opatch resource.

**Comment by userMar2017** [comment]
Sorry for the delayed response. These arguments are provided by opatch.
